analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/ObjectConstraint.cs

```diff
@@ -30,6 +32,9 @@ internal sealed class ObjectConstraint : SymbolicConstraint
             // x == "" ? <NotNull> : <unknown, could be Null or NotNull here>
             this == Null ? NotNull : null;
 
+        public override bool PreserveOnFieldReset(IFieldSymbol field) =>
+            field.IsStatic;
```

Comment:
It seemed more natural to me to let the constraint itself decide, what kind of fields should be preserved.

Reply:
No, we need to remove all of them. There's not much difference between modified static field or modified instance field. Any of them can be modified by a method invocation.